### ISSUE TITLE
IllegalStateException in laika:copyPDF while executing laika:site without PDF

### ISSUE THREAD

**Issue by pcmehlitz** [planet42/Laika#32]
when setting an excludeFilter (that just prunes "attic" directories), I get the attached IllegalStateException. I run laika:site, the exception gets reported in laika:copyPDF (although includePDF is set to false).

The html output gets generated, including the "installation" directory it complains about. The excludeFilter works, it prunes all html under attic/

The exception goes away if I don't set the excludeFilter.

java.lang.IllegalStateException: Unable to create directory .../target/docs/site/installation
        at laika.io.OutputProvider$DirectoryOutputProvider.newChild(OutputProvider.scala:80)
        at laika.api.Render$$anonfun$laika$api$Render$$collectOperations$1$4.apply(Render.scala:192)
        at laika.api.Render$$anonfun$laika$api$Render$$collectOperations$1$4.apply(Render.scala:192)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.immutable.List.foreach(List.scala:318)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
        at scala.collection.AbstractTraversable.map(Traversable.scala:105)
        at laika.api.Render.laika$api$Render$$collectOperations$1(Render.scala:192)
        at laika.api.Render.render(Render.scala:199)
        at laika.api.Render$RenderMappedOutput$$anon$2.renderTo(Render.scala:366)
        at laika.api.Render$MappedTreeTarget$class.toTree(Render.scala:333)
        at laika.api.Render$RenderMappedOutput$$anon$2.toTree(Render.scala:365)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59$$anonfun$60.apply(LaikaSbtPlugin.scala:306)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59$$anonfun$60.apply(LaikaSbtPlugin.scala:295)
        at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:235)
        at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:235)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:249)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:245)
        at sbt.Difference.apply(Tracked.scala:224)
        at sbt.Difference.apply(Tracked.scala:206)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:245)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:244)
        at sbt.Difference.apply(Tracked.scala:224)
        at sbt.Difference.apply(Tracked.scala:200)
        at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:244)
        at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:242)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59.apply(LaikaSbtPlugin.scala:353)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59.apply(LaikaSbtPlugin.scala:293)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.immutable.List.foreach(List.scala:318)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
        at scala.collection.AbstractTraversable.map(Traversable.scala:105)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6.apply(LaikaSbtPlugin.scala:293)
        at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6.apply(LaikaSbtPlugin.scala:269)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
        at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
        at sbt.std.Transform$$anon$4.work(System.scala:63)

 ...

**Comment by jenshalm** [comment]
Could you confirm whether you are using the 0.6 release or an earlier snapshot and provide the full sbt build to reproduce, including your input directory layout?

**Comment by pcmehlitz** [comment]
yes, I'm using 0.6.

The effect is spurious, it seems to be a threading issue.

The snapshot is going to take a couple days since this is a NASA project for which I need to go through some release process. However, there is not much Laika specific config:

  val laikaSettings = LaikaPlugin.defaults ++ Seq(
    sourceDirectories in Laika := Seq(file("doc")),
    rawContent in Laika := true,
    includePDF in Laika := false,
    includeAPI in Laika := false,
    excludeFilter in Laika := new FileFilter {
      override def accept(file:File): Boolean = file.getName == "attic"
    },
    aggregate in Laika := false
  )


The input directory layout is as follows:

~/projects/race/wow/doc$ tree
.
├── about.md
├── attic
│   ├── caveats.md
│   ├── ide.md
│   ├── learning.md
│   ├── libraries.md
│   ├── log.md
│   └── todo.md
├── default.template.html
├── design
│   ├── design-principles.md
│   ├── directory.conf
│   └── directory-structure.md
├── directory.conf
├── glossary.md
├── graphics
│   └── race128.png
├── installation
│   ├── build.md
│   ├── directory.conf
│   ├── download.md
│   ├── osx.md
│   └── prerequisites.md
├── race.css
├── specs
│   └── nonfunctional.md
└── usage
    ├── configuration.md
    ├── directory.conf
    ├── encryption.md
    ├── examples.md
    └── running.md


-- Peter

**Comment by jenshalm** [comment]
Ok, thanks a lot, I'll try to reproduce in the coming days!

**Comment by pcmehlitz** [comment]
thanks, I’ll let you know when the release is done. I can reproduce this fairly consistently on every 2nd or 3rd time I run Laika. This is on Linux with JDK 8u92, I’ll try OS X in a bit.

I’m planing to use Laika in a critical role for what I call unified repository - sources, documentation, presentation and issues all within one repository, for which Laika’s extensibility is key. In that context I have a couple of questions:

(1) what is the best way to refer to global resources (*.css, images etc.) from nested directories, short of duplicating templates? I currently use the directory.confs to add a “rootPath” setting which is then used in my top-level template (e.g. “..href=“{{rootPath}}/race.css”.. ”), but it seems this should be a system variable since its a general problem.

(2) is there any work planned on renderers for presentation (slides) or publication (LaTeX)?

(3) can I run Laika on disparate roots with different content, say on doc/ to generate an HTML site, and on publication/ to generate papers?

— Peter

> Ok, thanks a lot, I'll try to reproduce in the coming days!

**Comment by jenshalm** [comment]
(1) This scenario is well supported. The setting for input directories is `sourceDirectories: Seq[File]`, so you can specify, say, the root directory for a set of markup files, plus a shared directory with styles and images, in a way that Laika will see them as merged to a single root without the need for you to actually copy and duplicate the shared files. Internally Laika operates with a virtual tree, so not everything has to directly map to the file system. You can also programmatically add resources if required.

(2) Not pre-1.0, meaning not in the coming 18 months or so. After 1.0 the feature set mostly depends on feedback and requests, so it would be an option then. Before that contributions would be welcome, too, of course. So far there hasn't been anything apart from tiny fixes, but I'd be open to change that, I'd only need to set up a few contribution guidelines in that case.

(3) I'm not sure I understand this question correctly. Do you mean you want to run two completely independent operations with different inputs, config and output with a single command from within the same sbt build? I guess in this case you'd need to work with sbt configuration scopes and run laika:site twice with different configs.

**Comment by pcmehlitz** [comment]
> (1) This scenario is well supported. The setting for input directories is sourceDirectories: Seq[File], so you can specify, say, the root directory for a set of markup files, plus a shared directory with styles and images, in a way that Laika will see them as merged to a single root without the need for you to actually copy and duplicate the shared files. Internally Laika operates with a virtual tree, so not everything has to directly map to the file system. You can also programmatically add resources if required.

cool, that is all I need.

> (2) Not pre-1.0, meaning not in the coming 18 months or so. After 1.0 the feature set mostly depends on feedback and requests, so it would be an option then. Before that contributions would be welcome, too, of course. So far there hasn't been anything apart from tiny fixes, but I'd be open to change that, I'd only need to set up a few contribution guidelines in that case.

it depends on how RACE is evolving, but I might come up with some contributions. Slides are probably first.

> (3) I'm not sure I understand this question correctly. Do you mean you want to run two completely independent operations with different inputs, config and output with a single command from within the same sbt build?

I mean different task invocations, but with sourceDirectories and renderers as argument. I can wrap this into my own tasks though

Thanks,
-- Peter

**Comment by jenshalm** [comment]
Renderers as arguments are actually possible, with the `generate` task, e.g. `generate html pdf`. But `sourceDirectories` is only available as a setting. There are so many settings that I wasn't sure whether it makes sense to pull some of them out as arguments (and which).

**Comment by pcmehlitz** [comment]
I saw, but I think in many cases (e.g. slides) the content is going to be rendering specific. However, it's easy enough to write a little project specific task that overrides the Laika settings

I saw that you ignore *.rst image rendering options. Is that in the queue? Otherwise I just use embedded HTML since styles are not really suitable (differs for most images)

-- Peter

**Comment by jenshalm** [comment]
Feedback volume regarding Laika has been low to moderate so far, that's why I am not aware of many possible use cases. One I had in mind was that of projects like Akka, where the same set of input files is indeed used to produce all the output formats (HTML, PDF, epub).

Writing custom tasks might be the most sensible choice in your case, because starting to expose more settings as arguments might only bring half of the solution (people might find out they need even more settings to be different per set of input files).

Like I hinted at, working with custom configuration scopes might also work, I don't have an example right now, but I might try to come up with one later.

Regarding rst image rendering options, that seems to be a reasonable addition, feel free to create a new issue for that.

**Comment by jenshalm** [comment]
I'm afraid I will have to close this ticket with "cannot reproduce". I tried two things, first reproducing the issue by running the task myself, but since this is very likely related to a race condition it is not trivial to reproduce. Then I analysed the code to see if I can spot any potential open doors for race conditions, but did not see any unless other tools are running in parallel. The exception you have shown can only occur if a file operation happens between the check for the existence of a file and the subsequent operation, which is something where I cannot see how this can be caused by Laika itself.

You can reopen anytime if you find more hints as to what exactly is causing this behaviour.

**Comment by pcmehlitz** [comment]
I still get the exceptions fairly often, although not 100% reproducible (OS X 10.10.5, SBT 0.13.13, Java 1.8.0_91)

To reproduce:

  git clone https://github.com/NASARace/race.git
  cd race
  sbt
  [sbt] mkManual

The latest log is appended. Please note it is about a directory within SBTs realm (target/doc/installation). Unless I miss my guess there is a race going on between SBT and Laika. It always happens within race/laika:copyPDF and I don’t produce PDF.

It is not a critical bug though - the html output is produced correctly.

— Peter

[race] >> mkManual
[info] Set current project to race (in build file:/Users/pmehlitz/projects/race/race/)
[info] Reading files from doc/manual
[info] Reading files from doc/manual
[info] Parsing 28 markup documents, 1 template, 5 configurations ...
[info] Parsing 28 markup documents, 1 template, 5 configurations ...
[info] Rendering 28 HTML documents, copying 27 static files ...
[info] Generated html in /Users/pmehlitz/projects/race/race/target/doc
[trace] Stack trace suppressed: run last race/laika:copyPDF for the full output.
[error] (race/laika:copyPDF) java.lang.IllegalStateException: Unable to create directory /Users/pmehlitz/projects/race/race/target/doc/installation
[race] >> last race/laika:copyPDF
java.lang.IllegalStateException: Unable to create directory /Users/pmehlitz/projects/race/race/target/doc/installation
	at laika.io.OutputProvider$DirectoryOutputProvider.newChild(OutputProvider.scala:80)
	at laika.api.Render$$anonfun$laika$api$Render$$collectOperations$1$4.apply(Render.scala:192)
	at laika.api.Render$$anonfun$laika$api$Render$$collectOperations$1$4.apply(Render.scala:192)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at laika.api.Render.laika$api$Render$$collectOperations$1(Render.scala:192)
	at laika.api.Render.render(Render.scala:199)
	at laika.api.Render$RenderMappedOutput$$anon$2.renderTo(Render.scala:366)
	at laika.api.Render$MappedTreeTarget$class.toTree(Render.scala:333)
	at laika.api.Render$RenderMappedOutput$$anon$2.toTree(Render.scala:365)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59$$anonfun$60.apply(LaikaSbtPlugin.scala:306)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59$$anonfun$60.apply(LaikaSbtPlugin.scala:295)
	at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
	at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:267)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:263)
	at sbt.Difference.apply(Tracked.scala:224)
	at sbt.Difference.apply(Tracked.scala:206)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:263)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:262)
	at sbt.Difference.apply(Tracked.scala:224)
	at sbt.Difference.apply(Tracked.scala:200)
	at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:262)
	at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:260)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59.apply(LaikaSbtPlugin.scala:353)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6$$anonfun$59.apply(LaikaSbtPlugin.scala:293)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6.apply(LaikaSbtPlugin.scala:293)
	at laika.sbt.LaikaSbtPlugin$Tasks$$anonfun$57$$anonfun$apply$5$$anonfun$apply$6.apply(LaikaSbtPlugin.scala:269)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (race/laika:copyPDF) java.lang.IllegalStateException: Unable to create directory /Users/pmehlitz/projects/race/race/target/doc/installation

**Comment by jenshalm** [comment]
Ok, I'll give it another shot with your sample and see if I can reproduce it on my machine.

But there shouldn't be a race between Laika and sbt unless you point the target for your transformation to a directory also used by sbt. In theory the exception above can only occur if one or more Laika tasks get started twice, but from analysing the code I could not see how this could happen.

**Comment by jenshalm** [comment]
I was finally able to reproduce it, and immediately realised the root cause of the problem. The site task runs several dependent tasks twice which is not intended and the cause of the race condition.

Sample log output from running the task from your project:
```
[info] Reading files from doc/manual
[info] Reading files from doc/manual
[info] Parsing 28 markup documents, 1 template, 5 configurations ...
[info] Parsing 28 markup documents, 1 template, 5 configurations ...
```
These logs should not appear twice!

Now that the issue is narrowed down, I'll close this ticket and track the actual root problem in a new ticket: #44. Without running any tasks twice there won't be any race condition, so the exception should no longer occur.

Many thanks for the feedback... and the patience! :-)

**Comment by pcmehlitz** [comment]
thanks - glad to be of help.
BTW - Laika is used for both the official RACE website (http://nasarace.github.io/race/) and presentation slides (http://nasarace.github.io/race/slides/DASC16.html), although the latter could probably still use some refinement.

— Peter
